[SPARK-22303][SQL] Handle Oracle specific jdbc types in OracleDialect

taroplus opened the pull request:
TIMESTAMP (-101), BINARY_DOUBLE (101) and BINARY_FLOAT (100) are handled in OracleDialect

## What changes were proposed in this pull request?

When a oracle table contains columns whose type is BINARY_FLOAT or BINARY_DOUBLE, spark sql fails to load a table with SQLException

```
java.sql.SQLException: Unsupported type 101
 at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$getCatalystType(JdbcUtils.scala:235)
 at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$8.apply(JdbcUtils.scala:292)
 at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$8.apply(JdbcUtils.scala:292)
 at scala.Option.getOrElse(Option.scala:121)
 at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$.getSchema(JdbcUtils.scala:291)
 at org.apache.spark.sql.execution.datasources.jdbc.JDBCRDD$.resolveTable(JDBCRDD.scala:64)
 at org.apache.spark.sql.execution.datasources.jdbc.JDBCRelation.<init>(JDBCRelation.scala:113)
 at org.apache.spark.sql.execution.datasources.jdbc.JdbcRelationProvider.createRelation(JdbcRelationProvider.scala:47)
 at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:306)
 at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:178)
 at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:146)
```

## How was this patch tested?

I updated a UT which covers type conversion test for types (-101, 100, 101), on top of that I tested this change against actual table with those columns and it was able to read and write to the table.

taroplus committed:
Handle Oracle specific jdbc types in OracleDialect

TIMESTAMP (-101), BINARY_DOUBLE (101) and BINARY_FLOAT (100) are handled in OracleDialect

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #82951 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82951/testReport)** for PR 19548 at commit [`51c616c`](https://github.com/apache/spark/commit/51c616c6e501ad539f4e18cf604250a66edf1a2e).

gatorsmile commented:
Could you add a test case to `OracleIntegrationSuite.scala`?

Below is the instruction how to run the docker-based testsuite `OracleIntegrationSuite.scala `.

> build/mvn -Pyarn -Phadoop-2.6 -Dhadoop.version=2.6.0  -Phive-thriftserver -Phive -DskipTests  install
> Before running dockers test . You need to set the docker env variabled. Following does the magic:
> eval $(docker-machine env default)
> build/mvn -Pdocker-integration-tests -pl :spark-docker-integration-tests_2.11  compile test

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -28,25 +28,28 @@ private case object OracleDialect extends JdbcDialect {
+        }
+      case -101 => Some(TimestampType) // Value for Timestamp with Time Zone in Oracle
+      case 100 => Some(FloatType) // Value for OracleTypes.BINARY_FLOAT
+      case 101 => Some(DoubleType) // Value for OracleTypes.BINARY_DOUBLE
```
Question: Is the value range of `OracleTypes.BINARY_DOUBLE` identical to our Spark Double type?

Also, `OracleTypes.BINARY_FLOAT ` with Spark Float type

taroplus replied:
This should match java's double / float definition

SparkQA commented:
**[Test build #82951 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82951/testReport)** for PR 19548 at commit [`51c616c`](https://github.com/apache/spark/commit/51c616c6e501ad539f4e18cf604250a66edf1a2e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82951/
Test PASSed.

taroplus committed:
Integration test for oracle types BINARY_DOUBLE and BINARY_FLOAT

SparkQA commented:
**[Test build #82965 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82965/testReport)** for PR 19548 at commit [`91e911d`](https://github.com/apache/spark/commit/91e911d92f96c251de10a86a73f854b4993ede39).

taroplus commented:
Okay new integration test has been added

According to their [document](https://docs.oracle.com/database/121/TTSQL/types.htm#TTSQL148) 

> BINARY_DOUBLE is a 64-bit, double-precision, floating-point number.
> 
> BINARY_FLOAT is a 32-bit, single-precision, floating-point number.
> 
> Both BINARY_FLOAT and BINARY_DOUBLE support the special values Inf, -Inf, and NaN (not a number) and conform to the IEEE standard.

This should match java's double / float definition

srowen commented on external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -307,4 +315,32 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
+    val props = new Properties()
+
+    // write it back to the table (append mode)
+    val data = spark.sparkContext.parallelize(Seq(Row(1.1D, 2.2F)))
```
Nit: I think it's more conventional to write "1.1" and "2.2f"

taroplus replied:
updated

srowen commented on external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -108,6 +112,10 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
     conn.commit();
+
+
+    conn.prepareStatement("CREATE TABLE oracle_types (d BINARY_DOUBLE, f BINARY_FLOAT)").executeUpdate();
```
No semicolon at the end of lines

taroplus replied:
removed trailing semi-colons throughout this file

srowen commented on external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -50,9 +52,11 @@ import org.apache.spark.tags.DockerTest
 class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLContext {
   import testImplicits._
+  // To make === between double tolerate inexact values
+  implicit val doubleEquality = TolerantNumerics.tolerantDoubleEquality(0.01)
```
Curious why it would be that inequal somewhere in these tests?

taroplus replied:
removed this line and the test still passes

srowen commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -815,6 +815,12 @@ class JDBCSuite extends SparkFunSuite
       Some(DecimalType(DecimalType.MAX_PRECISION, 10)))
     assert(oracleDialect.getCatalystType(java.sql.Types.NUMERIC, "numeric", 0, null) ==
       Some(DecimalType(DecimalType.MAX_PRECISION, 10)))
+    assert(oracleDialect.getCatalystType(100, "BINARY_FLOAT", 0, null) ==
```
Let's define constants somewhere suitable for 100/100/-101

taroplus replied:
Made them constants defined in OracleDialect

srowen reviewed (changes requested):
Are other DB-specific types supported in Spark?

SparkQA commented:
**[Test build #82965 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82965/testReport)** for PR 19548 at commit [`91e911d`](https://github.com/apache/spark/commit/91e911d92f96c251de10a86a73f854b4993ede39).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82965/
Test PASSed.

taroplus committed:
Address review feedbacks

 - No trailing semi-colons
 - constatns for 100, 101 and -101
 - just compare double / float

SparkQA commented:
**[Test build #82968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82968/testReport)** for PR 19548 at commit [`28c7ce8`](https://github.com/apache/spark/commit/28c7ce8484d2e0182b213ebc63b857f047c06ca6).

SparkQA commented:
**[Test build #82968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82968/testReport)** for PR 19548 at commit [`28c7ce8`](https://github.com/apache/spark/commit/28c7ce8484d2e0182b213ebc63b857f047c06ca6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82968/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented:
Thanks! Merged to master.
